libcaf_core/caf/flow/op/sample.test.cpp

```diff
@@ -0,0 +1,355 @@
+
+constexpr auto fwd_ctrl = caf::flow::op::sample_emit_t{};
+
+struct fixture : test::fixture::deterministic, test::fixture::flow {
+  // Similar to sample::subscribe, but returns a sample_sub pointer instead of
+  // type-erasing it into a disposable.
+  template <class T = int>
+  auto raw_sub(caf::flow::observable<int> in,
+               caf::flow::observable<int64_t> select,
+               caf::flow::observer<int> out) {
+    using sub_t = caf::flow::op::sample_sub<T>;
+    auto ptr = make_counted<sub_t>(coordinator(), out);
+    ptr->init(in, select);
+    out.on_subscribe(caf::flow::subscription{ptr});
+    return ptr;
+  }
+
+  template <class T>
+  auto make_never_sub(caf::flow::observer<T> out) {
+    return make_counted<caf::flow::op::never_sub<T>>(coordinator(), out);
+  }
+};
+
+WITH_FIXTURE(fixture) {
+
+SCENARIO("the sample operator emits items at regular intervals") {
+  GIVEN("an observable") {
+    WHEN("calling .sample(1s)") {
+      THEN("the observer receives most recent values after 1s") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto expected = std::vector<int>{32, 64, 512};
+        auto closed = std::make_shared<bool>(false);
+        auto pub = caf::flow::multicaster<int>{coordinator()};
+        sys.spawn([&pub, outputs, closed](caf::event_based_actor* self) {
+          pub.as_observable()
+            .observe_on(self) //
+            .sample(1s)
+            .do_on_complete([closed] { *closed = true; })
+            .for_each([outputs](const int& xs) { outputs->emplace_back(xs); });
+        });
+        dispatch_messages();
+        print_debug("emit the first six items");
+        pub.push({1, 2, 4, 8, 16, 32});
+        run_flows();
+        dispatch_messages();
+        print_debug("force an empty sample");
+        advance_time(1s);
+        dispatch_messages();
+        print_debug("force a sample with a single element");
+        pub.push(64);
+        run_flows();
+        dispatch_messages();
+        advance_time(1s);
+        dispatch_messages();
+        print_debug("force an empty sample");
+        advance_time(1s);
+        dispatch_messages();
+        print_debug("emit the last items and close the source");
+        pub.push({128, 256, 512});
+        pub.close();
+        run_flows();
+        dispatch_messages();
+        advance_time(1s);
+        dispatch_messages();
+        check_eq(*outputs, expected);
+        check(*closed);
+      }
+    }
+  }
+}
+
+SCENARIO("the sample operator forwards errors") {
+  GIVEN("an observable that produces some values followed by an error") {
+    WHEN("calling .sample() on it") {
+      THEN("the observer receives the values and then the error") {
+        auto outputs = std::make_shared<std::vector<int>>();
+        auto err = std::make_shared<error>();
+        auto pub = caf::flow::multicaster<int>{coordinator()};
+        sys.spawn([&pub, outputs, err, this](caf::event_based_actor* self) {
```

Comment:
Why do we capture `this`?

Reply:
`this` was captured to call `make_observable`. This has been fixed by calling `self->make_observable()`.